File: 1.12/main.c
#include "../nob.h"

int main(int argc, char **argv) {

	FILE *input = open_input_or_die(argc, argv);

    char dir = '\0';
    int distance = -1;
    int state = 50;
    ll part1 = 0, part2 = 0;
    string line = NULL;
    size_t size = 0;
   	int tmp = -1;

	
    while (getline_or_die(&line, &size, input) != -1) {
    
    	dir = line[0];
    	distance = (int) strtol(line+1, NULL, 0);
    	
		tmp = dir == 'R' ? (100 - state) % 100 : state % 100;
        if (tmp == 0) tmp = 100;

		state = dir == 'R' ? (state + distance) % 100 : (state - distance) % 100; 
        if (state < 0) state += 100;
        if (state == 0) part1++;
        
		if (distance >= tmp) part2 += 1 + (distance - tmp) / 100;
    }


    printf("Part 1: %lld\n", part1);
    printf("Part 2: %lld\n", part2);

	free(line);
	fclose(input);
	
    return 0;
}

File: 1.12/main.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
	let args: Vec<String> = env::args().collect();
	
    if args.len() < 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }

    let file = File::open(&args[1]).unwrap();
    let reader = BufReader::new(file);
    let mut state: i32 = 50;
    let mut part1 = 0;
    let mut part2 = 0;
    
    for line in reader.lines() {
        let line = line.unwrap();
        
        if let Some(dir) = line.chars().next() {
            assert!(dir == 'L' || dir == 'R');
            if let Ok(distance) = line[1..].parse::<i32>() {

                let mut tmp = if dir == 'R' { 
                    (100 - state) % 100 
                } else { 
                    state % 100 
                };
                
                if tmp == 0 { 
                    tmp = 100; 
                }
                
                state = if dir == 'R' { 
                    (state + distance) % 100 
                } else { 
                    (state - distance) % 100 
                };

                if state < 0 { 
                    state += 100; 
                }
                
                if distance >= tmp { 
                    part2 += 1 + (distance - tmp) / 100; 
                }

                if state == 0 { 
                    part1 += 1; 
                }
            }
        }
    }
    
    println!("part1: {part1}");
    println!("part2: {part2}");
}
